### PR TITLE
perf: avoid temporary Vec allocations when computing commitments

### PR DESCRIPTION
**Pull request by mina86**
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests.
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 1]
```diff
@@ -86,16 +86,16 @@ pub(crate) fn compute_packet_commitment(
     timeout_height: &TimeoutHeight,
     timeout_timestamp: &Timestamp,
 ) -> PacketCommitment {
-    let mut hash_input = timeout_timestamp.nanoseconds().to_be_bytes().to_vec();
+    use sha2::Digest;
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/commitment.rs]:
```suggestion
```

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 2]
```diff
@@ -86,16 +86,16 @@ pub(crate) fn compute_packet_commitment(
...
 
-    let revision_height = timeout_height.commitment_revision_height().to_be_bytes();
-    hash_input.append(&mut revision_height.to_vec());
+    hash_input[..8].copy_from_slice(&timeout_timestamp.nanoseconds().to_be_bytes());
+    hash_input[8..16].copy_from_slice(&timeout_height.commitment_revision_number().to_be_bytes());
+    hash_input[16..24].copy_from_slice(&timeout_height.commitment_revision_height().to_be_bytes());
 
-    let packet_data_hash = hash(packet_data);
-    hash_input.append(&mut packet_data_hash.to_vec());
+    let packet_data_hash = sha2::Sha256::digest(packet_data);
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/commitment.rs]:
```suggestion
    let packet_data_hash = hash(packet_data);
```

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 3]
```diff
@@ -109,8 +109,32 @@ pub(crate) fn compute_ack_commitment(ack: &Acknowledgement) -> AcknowledgementCo
 ///
 /// Note that computing commitments with anything other than SHA256 will
 /// break the Merkle proofs of the IBC provable store.
-fn hash(data: impl AsRef<[u8]>) -> Vec<u8> {
+fn hash(data: &[u8]) -> Vec<u8> {
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/commitment.rs]: Given the above suggestions, this `hash` function should accordingly return: `[u8; 32]`

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 4]
```diff
@@ -109,8 +109,32 @@ pub(crate) fn compute_ack_commitment(ack: &Acknowledgement) -> AcknowledgementCo
...
     use sha2::Digest;
 
-    sha2::Sha256::digest(&data).to_vec()
+    sha2::Sha256::digest(data).to_vec()
+}
+
+#[test]
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/commitment.rs]: As for the tests, I’d like to have @rnbguy's input!

**Review by Farhad-Shabani — COMMENTED** [review]
Thanks @mina86! Left a few suggestions.
Also, for our context, appreciate let us know what led to this change? Have you established any kind of benchmarks or obtained a preliminary estimate of the impact this change will have on your project's performance?

**Comment by mina86** [comment]
No, this was purely from looking at the code. I’ve done no benchmarks on it.

The reason why I was looking at the code was because I needed to make sure commitments are always 32-byte long.

By the way, ideally I’d like commitment to be `struct PacketCommitment([u8; 32]);` but that’s a completely different conversation (or `struct PacketCommitment(Box<[u8; 32]>);`; point is to have it fixed-size).

File: crates/ibc/src/core/ics04_channel/commitment.rs (hunk 4 above)
Comment (mina86) [reply crates/ibc/src/core/ics04_channel/commitment.rs]: The test here was purely to make sure that the output doesn’t change during refactoring. I can split this PR into two (one adding the test and the other doing the refactoring) if that’d help to demonstrate it.

**Comment by Farhad-Shabani** [comment]
> No, this was purely from looking at the code. I’ve done no benchmarks on it.
> The reason why I was looking at the code was because I needed to make sure commitments are always 32-byte long.

Alright, Gotcha.

> By the way, ideally I’d like commitment to be `struct PacketCommitment([u8; 32]);` but that’s a completely different conversation (or `struct PacketCommitment(Box<[u8; 32]>);`; point is to have it fixed-size).

I see where you are coming from. Exactly, this is a totally different topic and a critical one. Better to surface it later when we have enough bandwidth to fully evaluate the changes around.

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 5]
```diff
@@ -86,31 +86,51 @@ pub(crate) fn compute_packet_commitment(
...
-    hash_input.append(&mut packet_data_hash.to_vec());
-
-    hash(&hash_input).into()
+    PacketCommitment(hash(&hash_input).into())
 }
 
 /// Compute the commitment for an acknowledgement.
 pub(crate) fn compute_ack_commitment(ack: &Acknowledgement) -> AcknowledgementCommitment {
-    hash(ack.as_ref()).into()
+    AcknowledgementCommitment(hash(ack.as_ref()).into())
```
Comment (rnbguy) [inline crates/ibc/src/core/ics04_channel/commitment.rs]:
```suggestion
    hash(ack.as_ref()).to_vec().into()
```

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 6]
```diff
@@ -86,31 +86,51 @@ pub(crate) fn compute_packet_commitment(
     timeout_height: &TimeoutHeight,
     timeout_timestamp: &Timestamp,
 ) -> PacketCommitment {
-    let mut hash_input = timeout_timestamp.nanoseconds().to_be_bytes().to_vec();
+    let mut hash_input = [0; 3 * 8 + 32];
```
Comment (rnbguy) [inline crates/ibc/src/core/ics04_channel/commitment.rs]:
```suggestion
    let mut hash_input = [0; 8 * 3 + 32];
```

Just being pedantic as it is 8 bytes x3 and 32 bytes x1.

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 7]
```diff
@@ -86,31 +86,51 @@ pub(crate) fn compute_packet_commitment(
...
+    hash_input[24..].copy_from_slice(&hash(packet_data));
 
-    let revision_height = timeout_height.commitment_revision_height().to_be_bytes();
-    hash_input.append(&mut revision_height.to_vec());
-
-    let packet_data_hash = hash(packet_data);
-    hash_input.append(&mut packet_data_hash.to_vec());
-
-    hash(&hash_input).into()
+    PacketCommitment(hash(&hash_input).into())
```
Comment (rnbguy) [inline crates/ibc/src/core/ics04_channel/commitment.rs]:
```suggestion
    hash(&hash_input).to_vec().into()
```

File: crates/ibc/src/core/ics04_channel/commitment.rs (hunk 4 above)
Comment (rnbguy) [reply crates/ibc/src/core/ics04_channel/commitment.rs]: @mina86 can you please put the tests under 

```rs
#[cfg(test)]
mod test { ... }
```

and, make two separate tests for `compute_packet_commitment` and `compute_ack_commitment`?

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 8]
```diff
@@ -86,31 +86,51 @@ pub(crate) fn compute_packet_commitment(
...
+        &Timestamp::from_nanoseconds(0x42).unwrap(),
+    );
+    assert_eq!(&want[..], got.as_ref());
+
+    // AcknowledgementCommitment
+    let want: [u8; 32] = [
+        5, 78, 222, 193, 208, 33, 31, 98, 79, 237, 12, 188, 169, 212, 249, 64, 11, 14, 73, 28, 67,
+        116, 42, 242, 197, 176, 171, 235, 240, 201, 144, 216,
+    ];
+    let ack = Acknowledgement::try_from(vec![0u8, 1, 2, 3]).unwrap();
```
Comment (rnbguy) [inline crates/ibc/src/core/ics04_channel/commitment.rs]: You may remove `u8` annotation here as `Acknowledgement` only `impl`s `TryFrom<Vec<u8>>`.

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 9]
```diff
@@ -86,31 +86,51 @@ pub(crate) fn compute_packet_commitment(
...
+}
+
+#[test]
+fn test_compute_commitment() {
+    // PacketCommitment
+    let want: [u8; 32] = [
+        169, 40, 181, 31, 98, 189, 84, 0, 145, 236, 69, 31, 78, 243, 69, 121, 79, 5, 158, 101, 145,
+        8, 22, 134, 97, 38, 220, 54, 79, 132, 204, 21,
+    ];
+    let got = compute_packet_commitment(
```
Comment (rnbguy) [inline crates/ibc/src/core/ics04_channel/commitment.rs]: If you want to use `want` and `got` naming, please use the `expected` and `actual` naming convention.

File: .changelog/unreleased/improvements/939-less-alloc.md [hunk 10]
```diff
@@ -0,0 +1,2 @@
+- Reduce amount of temporary vector allocations.
```
Comment (rnbguy) [inline .changelog/unreleased/improvements/939-less-alloc.md]:
```suggestion
- Reduce vector allocations in Commitment computation. 
```

File: crates/ibc/src/core/ics04_channel/commitment.rs [hunk 11]
```diff
@@ -86,31 +86,51 @@ pub(crate) fn compute_packet_commitment(
...
 
-    sha2::Sha256::digest(&data).to_vec()
+    sha2::Sha256::digest(data).into()
+}
+
+#[test]
+fn test_compute_commitment() {
+    // PacketCommitment
+    let want: [u8; 32] = [
+        169, 40, 181, 31, 98, 189, 84, 0, 145, 236, 69, 31, 78, 243, 69, 121, 79, 5, 158, 101, 145,
```
Comment (rnbguy) [inline crates/ibc/src/core/ics04_channel/commitment.rs]: Can you please write these values in hex?
```rs
let want: [u8; 32] = [0xa9, ...]
```

**Review by rnbguy — CHANGES_REQUESTED** [review]
Great PR, @mina86 ! :pray: Just requested some changes.

**Review by rnbguy — APPROVED** [review]
Awesome! Thanks @mina86 :handshake:

**Pull request merged**